### PR TITLE
[HttpKernel][Framework] Locale aware services

### PR DESCRIPTION
**Pull request by neghmurken**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no 
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Added a `LocaleAwareInterface` (and also a new service tag `kernel.locale_aware`) to be implemented on services that require the current locale at request time.
Also, refactored the actual Translator service to implement the overmentioned interface

Todo : 

* [ ] Documention PR (will be written after some feedback)

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml [hunk 1]
```diff
@@ -147,8 +148,8 @@
             <argument type="service" id="Psr\Container\ContainerInterface" />
         </service>
 
-        <service id="translator_listener" class="Symfony\Component\HttpKernel\EventListener\TranslatorListener">
-            <argument type="service" id="translator" />
+        <service id="locale_aware_listener" class="Symfony\Component\HttpKernel\EventListener\LocaleAwareListener">
+            <argument /> <!-- locale aware services -->
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: `type="tagged" tag="locale_aware"` would also work isnt it

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Register all services that implements LocaleAwareInterface into the listener
+ *
+ * @author Pierre Bobiet <pierrebobiet@gmail.com>
+ */
+class RegisterLocaleAwareServicesPass implements CompilerPassInterface
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php]: if so, this would not be needed

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ */
+class LocaleAwareListener implements EventSubscriberInterface
+{
+    /**
+     * @var \Symfony\Component\Translation\LocaleAwareInterface[]
+     */
+    private $localeAwareServices;
+    private $requestStack;
+
+    public function __construct(\Traversable $localeAwareServices, RequestStack $requestStack)
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: `iterable $localeAwareServices`

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        $this->setLocaleAwareServices($localeAwareServices);
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: `$this->localeAwareServices = $localeAwareServices;` is fine IMHO with a `@param LocaleAwareInterface[] $localeAwareServices` in phpdoc

File: src/Symfony/Component/Translation/LocaleAwareInterface.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+namespace Symfony\Component\Translation;
+
+/**
+ * LocaleAwareInterface should be implemented by classes that needs the current locale
+ *
+ * @author Pierre Bobiet <pierrebobiet@gmail.com>
+ */
+interface LocaleAwareInterface
```
Comment (ro0NL) [inline src/Symfony/Component/Translation/LocaleAwareInterface.php]: im not sure this belongs here, what about a new contract `Contract\Service\LocalizedInterface`?

**Review by ro0NL — COMMENTED** [review]
i think this is a nice approach/feature :)

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        $this->requestStack = $requestStack;
+    }
+
+    private function setLocaleAwareServices(\Traversable $localeAwareServices)
+    {
+        foreach ($localeAwareServices as $i => $service) {
```
Comment (jvasseur) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: This means you will lose the laziness of iterable arguments, and force instantiation of all locale aware services as soon as the listener is instantiated.

File: src/Symfony/Component/Translation/LocaleAwareInterface.php (hunk 5 above)
Comment (neghmurken) [reply src/Symfony/Component/Translation/LocaleAwareInterface.php]: I'll do that :)

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php (hunk 6 above)
Comment (jvasseur) [reply src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: Maybe the check could be done in a compiler pass like it's done for voters.

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * Inject request locale to each services that implements LocaleAwareInterface.
+ *
+ * @author Pierre Bobiet <pierrebobiet@gmail.com>
+ */
+class LocaleAwareListener implements EventSubscriberInterface
+{
+    /**
+     * @var \Symfony\Contracts\Service\LocaleAwareInterface[]
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: usually we put it on the constructor's `@param` notation, as the property type can be inferred from that. Also namespace does not need to be included, this interface is already in "use"

File: src/Symfony/Component/HttpKernel/Tests/EventListener/LocaleAwareListenerTest.php [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+class LocaleAwareListenerTest extends TestCase
+{
+    private $listener;
+    private $localeAwareService;
+    private $requestStack;
+
+    protected function setUp()
+    {
+        $this->localeAwareService = $this->getMockBuilder(LocaleAwareInterface::class)->getMock();
+        $this->requestStack = $this->getMockBuilder('Symfony\Component\HttpFoundation\RequestStack')->getMock();
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/Tests/EventListener/LocaleAwareListenerTest.php]: `RequestStack::class`

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                    $i,
+                    __METHOD__,
+                    LocaleAwareInterface::class,
+                    \is_object($service) ? \get_class($service) : \gettype($service)
+                ));
+            }
+
+            try {
+                $service->setLocale($request->getLocale());
+            } catch (\InvalidArgumentException $e) {
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: this should be documented in the contract, but before we continue let's ask @nicolas-grekas if we're on the right track.

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            KernelEvents::FINISH_REQUEST => array(array('onKernelFinishRequest', 0)),
+        );
+    }
+
+    private function setLocale(Request $request): void
+    {
+        foreach ($this->localeAwareServices as $i => $service) {
+            if (!$service instanceof LocaleAwareInterface) {
+                throw new TypeError(sprintf(
+                    'Value #%d of argument 1 passed to %s must be an instance of %s, %s given',
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: locale aware services are not an argument of `__METHOD__` here.

something like `Service must be an instance of "%s", got "%s".` would be more accurate :) but perhaps we should find a way to move the check to compile time, e.g.

```xml
<argument type="tagged" tag="locale_aware" instanceof="SF\...\LocaleAwareInterface"/>
```

**Review by ro0NL — COMMENTED** [review]
the autoconfiguration should also be updated (each service being an instance of LocaleAwareInterface should get the `locale_aware` tag automatically.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml [hunk 11]
```diff
@@ -147,8 +148,8 @@
             <argument type="service" id="Psr\Container\ContainerInterface" />
         </service>
 
-        <service id="translator_listener" class="Symfony\Component\HttpKernel\EventListener\TranslatorListener">
-            <argument type="service" id="translator" />
+        <service id="locale_aware_listener" class="Symfony\Component\HttpKernel\EventListener\LocaleAwareListener">
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: this service should not be defined in `translation.xml`, as it may be useful even when the translation component is not used in the project (as there might be other local-aware services).

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml [hunk 12]
```diff
@@ -147,8 +148,8 @@
...
+            <argument type="tagged" tag="locale_aware"/>
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: using a dedicated compiler pass rather than using `tagged` could allow to unregister the listener in case it has no locale aware services at all (which becomes relevant once the service is defined even when the translator is not configured, as it allows cases where we don't have locale aware services in the core).

File: src/Symfony/Contracts/Service/LocaleAwareInterface.php [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Contracts\Service;
```
Comment (stof) [inline src/Symfony/Contracts/Service/LocaleAwareInterface.php]: should it actually be in the Service namespace here ? It is not specific to a service container AFAICT.

File: src/Symfony/Component/HttpKernel/EventListener/TranslatorListener.php [hunk 14]
```diff
@@ -24,6 +24,8 @@
  * Synchronizes the locale between the request and the translator.
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @deprecated since Symfony 4.2
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/EventListener/TranslatorListener.php]: this is missing the `@trigger_error`. And it should be 4.3 as we are in feature freeze

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\Event\FinishRequestEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Contracts\Service\LocaleAwareInterface;
+use TypeError;
+
+/**
+ * Inject request locale to each services that implements LocaleAwareInterface.
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: service

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+        $this->setLocale($parentRequest);
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return array(
+            // must be registered after the Locale listener
+            KernelEvents::REQUEST => array(array('onKernelRequest', 10)),
+            KernelEvents::FINISH_REQUEST => array(array('onKernelFinishRequest', 0)),
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]:
```php
KernelEvents::FINISH_REQUEST => 'onKernelFinishRequest',
```

File: src/Symfony/Component/HttpKernel/Tests/EventListener/LocaleAwareListenerTest.php (hunk 8 above)
Comment (xabbuh) [reply src/Symfony/Component/HttpKernel/Tests/EventListener/LocaleAwareListenerTest.php]: Actually, there is no reason to mock the request stack. Let's just use `new RequestStack()` here.

File: src/Symfony/Component/HttpKernel/Tests/EventListener/LocaleAwareListenerTest.php [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+    private function createRequest($locale)
+    {
+        $request = new Request();
+        $request->setLocale($locale);
+
+        return $request;
+    }
+
+    private function setMasterRequest($request)
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/Tests/EventListener/LocaleAwareListenerTest.php]: This method is not necessary when we don't mock the `RequestStack`. It can then be replaced with something like this:

```php
$this->requestStack->push($request);
```

Just make sure to push the subrequests too.

**Review by xabbuh — CHANGES_REQUESTED** [review]

File: src/Symfony/Contracts/Service/LocaleAwareInterface.php (hunk 13 above)
Comment (neghmurken) [reply src/Symfony/Contracts/Service/LocaleAwareInterface.php]: This namespace always has a `ResetInterface` which provides a utilitary functionality for services. So I thought `LocaleAwareInterface` (which also provides a functionality in the same vein) should be here. Any ideas?

**Comment by ro0NL** [comment]
curious.. what about using the existing `LocaleListener` for this? Then we can use `tagged` again maybe?

cc @stof

File: src/Symfony/Contracts/Service/LocaleAwareInterface.php (hunk 13 above)
Comment (ro0NL) [reply src/Symfony/Contracts/Service/LocaleAwareInterface.php]: im not sure `Service` is NS is bound to DI/containers per se. I think we can put any service descriptor here..

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+ * Register all services that implements LocaleAwareInterface into the listener.
+ *
+ * @author Pierre Bobiet <pierrebobiet@gmail.com>
+ */
+class RegisterLocaleAwareServicesPass implements CompilerPassInterface
+{
+    private $listenerServiceId;
+    private $localeAwareTag;
+
+    public function __construct(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php]: the constructor should be declared on one line

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+            return;
+        }
+
+        $services = array();
+
+        foreach ($container->findTaggedServiceIds($this->localeAwareTag) as $id => $tags) {
+            $definition = $container->getDefinition($id);
+            $serviceClass = $container->getParameterBag()->resolveValue($definition->getClass());
+
+            if (!is_a($serviceClass, LocaleAwareInterface::class, true)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php]: this check is not needed - it should be removed to prevent forcing unneeded classes at compile time

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+    public static function getSubscribedEvents()
+    {
+        return array(
+            // must be registered after the Locale listener
+            KernelEvents::REQUEST => array(array('onKernelRequest', 10)),
+            KernelEvents::FINISH_REQUEST => 'onKernelFinishRequest',
+        );
+    }
+
+    private function setLocale(Request $request): void
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: As it is implemented right now, a `LocaleAwareListener` cannot be reused in another context than as a listener.
But if we adopt this PR, I'm sure there might be more use cases where we'll need to "sync" the locale.
One example is when handling a message on the receiver side of a message bus.
How do we want to deal with that?

File: src/Symfony/Contracts/Service/LocaleAwareInterface.php [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+ * @author Pierre Bobiet <pierrebobiet@gmail.com>
+ */
+interface LocaleAwareInterface
+{
+    /**
+     * Sets the current locale.
+     *
+     * @param string $locale
+     */
+    public function setLocale($locale);
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/Service/LocaleAwareInterface.php]: Right now, I'm :-1: to add a setter in a contract.
That's a pure marker interface that encourages building mutable services.
We should seek for another design IMHO.

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
I'm mixed here:
- on one side this is very pragmatic;
- on the other side, this is building on top of service mutation - while services should be stateless.

I'd very prefer we could think about a stateless approach instead.

**Comment by nicolas-grekas** [comment]
I completely swapped my mind here and actually, we merged `LocaleAwareInterface` in Contracts v1.0.2 as a bug fix, see https://github.com/symfony/symfony/pull/29461

Would you mind rebasing, please?

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 22]
```diff
@@ -22,6 +22,7 @@ CHANGELOG
  * Deprecated `CachePoolPrunerPass`. Use `Symfony\Component\Cache\DependencyInjection\CachePoolPrunerPass` instead.
  * Deprecated support for legacy translations directories `src/Resources/translations/` and `src/Resources/<BundleName>/translations/`, use `translations/` instead.
  * Deprecated support for the legacy directory structure in `translation:update` and `debug:translation` commands.
+ * Added the `RegisterLocaleAwareServicesPass` and configured the `LocaleAwareListener`
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: should be moved to a new 4.3.0 section

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml [hunk 23]
```diff
@@ -73,5 +73,11 @@
         </service>
 
         <service id="services_resetter" class="Symfony\Component\HttpKernel\DependencyInjection\ServicesResetter" public="true" />
+
+        <service id="locale_aware_listener" class="Symfony\Component\HttpKernel\EventListener\LocaleAwareListener">
+            <argument /> <!-- locale aware services -->
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml]: type="collection"

File: src/Symfony/Component/HttpKernel/CHANGELOG.md [hunk 24]
```diff
@@ -18,6 +18,8 @@ CHANGELOG
  * Added support for using `service::method` to reference controllers, making it consistent with other cases. It is recommended over the `service:action` syntax with a single colon, which will be deprecated in the future.
  * Added the ability to profile individual argument value resolvers via the
    `Symfony\Component\HttpKernel\Controller\ArgumentResolver\TraceableValueResolver`
+ * Deprecated `TranslatorListener` in favor of `LocaleAwareListener`
+ * Added the registration of all `LocaleAwareInterface` implementations into the `LocaleAwareListener`
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/CHANGELOG.md]: for 4.3.0

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Symfony\Component\HttpKernel\DependencyInjection;
+
+use Symfony\Component\DependencyInjection\Argument\IteratorArgument;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Register all services that implements LocaleAwareInterface into the listener.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php]: that have the "locale_aware" tag

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+
+    public function __construct(string $listenerServiceId = 'locale_aware_listener', string $localeAwareTag = 'locale_aware')
+    {
+        $this->listenerServiceId = $listenerServiceId;
+        $this->localeAwareTag = $localeAwareTag;
+    }
+
+    public function process(ContainerBuilder $container)
+    {
+        if (false === $container->hasDefinition($this->listenerServiceId)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php]: `if (!$container->...`

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+            return;
+        }
+
+        $services = array();
+
+        foreach ($container->findTaggedServiceIds($this->localeAwareTag) as $id => $tags) {
+            $services[] = new Reference($id);
+        }
+
+        if (0 === \count($services)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php]: `if (!$services)`

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 28]
```diff
@@ -0,0 +1,74 @@
...
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\Event\FinishRequestEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Contracts\Translation\LocaleAwareInterface;
+
+/**
+ * Inject request locale to each service that implements LocaleAwareInterface.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: Pass the current locale to the provided services.

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+
+        $this->setLocale($parentRequest);
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return array(
+            // must be registered after the Locale listener
+            KernelEvents::REQUEST => array(array('onKernelRequest', 10)),
+            KernelEvents::FINISH_REQUEST => 'onKernelFinishRequest',
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: should it be a bit later?

File: src/Symfony/Component/HttpKernel/EventListener/TranslatorListener.php [hunk 30]
```diff
@@ -24,6 +24,8 @@
  * Synchronizes the locale between the request and the translator.
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @deprecated since Symfony 4.3
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/EventListener/TranslatorListener.php]: a runtime deprecation notice should also be added on top of the file - after the namespace declaration (see existing similar deprecations in the code base)

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 31]
```diff
@@ -0,0 +1,74 @@
...
+
+    public function onKernelRequest(GetResponseEvent $event): void
+    {
+        $this->setLocale($event->getRequest());
+    }
+
+    public function onKernelFinishRequest(FinishRequestEvent $event): void
+    {
+        if (null === $parentRequest = $this->requestStack->getParentRequest()) {
+            return;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: we should set the default locale here - so that the services are ready for another master request without leaking the locale from the previous one

File: src/Symfony/Contracts/CHANGELOG.md [hunk 32]
```diff
@@ -10,3 +10,4 @@ CHANGELOG
  * added `Service\ServiceSubscriberInterface` to declare the dependencies of a class that consumes a service locator
  * added `Service\ServiceSubscriberTrait` to implement `Service\ServiceSubscriberInterface` using methods' return types
  * added `Service\ServiceLocatorTrait` to help implement PSR-11 service locators
+ * added `Service\LocaleAwareInterface` to provide services a way to get the current locale at request time
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/CHANGELOG.md]: should be removed (unrelated to this PR)

**Review by nicolas-grekas — COMMENTED** [review]
thanks, here are some comments to help move forward.

**Comment by nicolas-grekas** [comment]
Small rebase needed + lowest version of httpkernel should be bumped in framework bundle to make tests pass.
Goot to me after that!

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        if (!$services) {
+            $container->removeDefinition($this->listenerServiceId);
+
+            return;
+        }
+
+        $container
+            ->findDefinition($this->listenerServiceId)
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php]: We can use `getDefinition()` here.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml [hunk 34]
```diff
@@ -140,11 +141,5 @@
             <tag name="kernel.cache_warmer" />
             <argument type="service" id="Psr\Container\ContainerInterface" />
         </service>
-
-        <service id="translator_listener" class="Symfony\Component\HttpKernel\EventListener\TranslatorListener">
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: Should we deprecate the service instead in case someone is relying on its presence?

Comment (neghmurken) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: How can we deprecate without it conflicting with the new `LocaleAwareListener` ?

Comment (xabbuh) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: Does it do any harm if both listeners are executed?

Comment (neghmurken) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: Yeah you're right. There is no undesirable side effect beside setting the locale twice

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 35]
```diff
@@ -7,6 +7,9 @@ CHANGELOG
  * Not passing the project directory to the constructor of the `AssetsInstallCommand` is deprecated. This argument will
    be mandatory in 5.0.
  * Deprecated the "Psr\SimpleCache\CacheInterface" / "cache.app.simple" service, use "Symfony\Contracts\Cache\CacheInterface" / "cache.app" instead
+ * Added an `help_html` form option to display the `help` text as HTML
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: this change doesn't look related

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 36]
```diff
@@ -7,6 +7,9 @@ CHANGELOG
...
+ * Added the `RegisterLocaleAwareServicesPass` and configured the `LocaleAwareListener`
+
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: the blank line should be removed

File: src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php [hunk 37]
```diff
@@ -0,0 +1,76 @@
...
+        return [
+            // must be registered after the Locale listener
+            KernelEvents::REQUEST => [['onKernelRequest', 15]],
+            KernelEvents::FINISH_REQUEST => 'onKernelFinishRequest',
+        ];
+    }
+
+    private function setLocale(string $locale, ?string $defaultLocale = null): void
+    {
+        foreach ($this->localeAwareServices as $i => $service) {
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php]: `$i` doesn't look needed here

File: src/Symfony/Component/HttpKernel/EventListener/TranslatorListener.php [hunk 38]
```diff
@@ -11,6 +11,8 @@
 
 namespace Symfony\Component\HttpKernel\EventListener;
 
+@trigger_error(sprintf('The "%s" class is deprecated since Symfony 4.3, use LocaleAwareListener instead.', TranslatorListener::class), E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/EventListener/TranslatorListener.php]: [...] since Symfony 4.3 and will be removed in 5.0. [...]

File: src/Symfony/Component/HttpKernel/Tests/DependencyInjection/RegisterLocaleAwareServicesPassTest.php [hunk 39]
```diff
@@ -0,0 +1,62 @@
...
+                    1 => new Reference('another_locale_aware_service'),
+                ]),
+                null,
+            ],
+            $container->getDefinition('locale_aware_listener')->getArguments()
+        );
+    }
+
+    /**
+     * @expectedException \Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/Tests/DependencyInjection/RegisterLocaleAwareServicesPassTest.php]: I would add an assertion like the following instead:

```php
$this->assertFalse($container->hasDefinition('locale_aware_listener'));
```

File: src/Symfony/Component/HttpKernel/Tests/EventListener/LocaleAwareListenerTest.php [hunk 40]
```diff
@@ -0,0 +1,120 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Tests\EventListener;
+
+use ArrayIterator;
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/Tests/EventListener/LocaleAwareListenerTest.php]: we don't add `use` statements for classes from the global namespace

**Review by xabbuh — COMMENTED** [review]
looks good apart from a few remaining minor comments

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md (hunk 35 above)
Comment (neghmurken) [reply src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: Bad conflict resolution while rebasing, my bad

**Review by nicolas-grekas — APPROVED** [review]
Green. I rebased this PR on latest master and made some tweaks to polish it. I removed the `translator_listener` service completely as keeping it would trigger deprecation notices. I don't think that's an issue.

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/DependencyInjection/RegisterLocaleAwareServicesPass.php]: It is needed to remove the listener if no locale aware services are found.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml (hunk 34 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml]: I removed the `translator_listener` service completely as keeping it would trigger deprecation notices. I don't think that's an issue.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @neghmurken.

**Comment by winzou** [comment]
Hello @neghmurken, @nicolas-grekas 
I'm not sure to understand this block in the `onKernelFinishRequest` method:

        if (null === $parentRequest = $this->requestStack->getParentRequest()) {
            $this->setLocale($event->getRequest()->getDefaultLocale());

            return;
        }

My usecase:
* I have a request with locale=`fr` and defaultLocale=`en`
* No subquery (hence a null `$parentRequest` in the above block)
* all strings translated in templates are done in `fr`, that's perfect
* but I also have strings translated in an event subscriber during the `kernel.terminate` event. These strings are translated in `en` => this is not what I want.

In the if I pasted, during the `kernel.finish` event, the `LocaleAwareListener` is defining the locale aware services (translator in my case) to the value of the request defaultLocale. Why? Why changing unilaterally the current locale of these services by the *default* locale of the request? If I define my *current* locale to `fr`, I want `fr` until the very end of the process, right?

This was not mentionned in the changelog and looks like a BC break (and a bug?) to me. Am I missing something?

Thanks for your insights!

**Comment by xabbuh** [comment]
@winzou Would you mind opening a new issue with your insights? Otherwise I am afraid that your input is likely to get lost.
